llvm/lib/Analysis/TargetTransformInfo.cpp

```diff
@@ -869,6 +870,18 @@ TargetTransformInfo::getOperandInfo(const Value *V) {
   return {OpInfo, OpProps};
 }
 
+InstructionCost TargetTransformInfo::getVecLibCallCost(
+    const int OpCode, const TargetLibraryInfo *TLI, VectorType *VecTy,
+    TTI::TargetCostKind CostKind) {
+  Type *ScalarTy = VecTy->getScalarType();
+  LibFunc Func;
+  if (TLI->getLibFunc(OpCode, ScalarTy, Func) &&
+      TLI->isFunctionVectorizable(TLI->getName(Func), VecTy->getElementCount()))
```

Comment:
This seems to be surreptitiously adding another mechanism to check for the presence of a vector math routine. Under what circumstance do you need to check for the cost of something that might not exist? I would expect TLI to be queried directly as part of a transformation and once they've concluded a vector math call exists then they'd simply query the cost of the call.

Reply:
Actually, in LoopVectorizer we have done it this way.

For the sake of reusing the code between the 2 vectorizers, I've put it in TTI, but after our discussion I've now moved it to `VectorUtils`. LoopVectorizer should now reuse this method.

Reply:
I disagree with this approach and still insist that it should be implemented in TTI rather doing some target specific "magic" in the vectorizers

Reply:
TTI should be costing known entities and not trying to second guess what a transformation is doing.  Whilst TLI provides a route for target specific queries relating to the availability of vector math routines, that doesn’t mean those mapping will be used.  The only source of truth is the transformation pass itself and thus it needs to ask the TTI the correct question based on this source of truth (i.e. whether it intends to use TLI mappings[1]). Both TTI and TLI abstract away the target specific queries relating to their design so I don’t understand what is the target specific “magic” you are worried about.

[1] This is very important because if the vector function is not used then a scalable vector FREM must have an invalid cost because there is no code generation support for it.

Reply:
It does not known what transsformation is doing, it does exactly what you said - estimates the cost of the target-specific code. WE're not emitting frem calls directly, we're still emitting regular frem vector code, which is then lowered specifically for ARM. And target specific TTI should be aware of this.

Reply:
OK, I see what you're saying.  I guess the problem here is that we're relying on a subsequent pass (ReplaceWithVecLib) to be present in the pipeline in order to correctly cost the FREM. With neither SLP nor TTI having this knowledge?

Reply:
Yes, and what I'm saying that TTI should handle this since this is target-specific lowering.
If you don't want to propagate TLI, then just copy the legality checks from TLI to TTI and the cost estimation for this specific lowering to the getArithmeticInstructionCost function. This is what we usually do in similar cases for other targets.